File: clusters.py
import networkx as nx
import random as r

G = nx.Graph()


def distance(x,y):  #to be defined
    pass


def pickNode(graph):    #picks a node, paper uses random, should we?
    return r.choice(graph.nodes)

#create a cluster around the center node, t is distance
def stoQuery(graph, t, center):
    Q = [center]    # weird way to implement queue quickly, list with index, high mem, so should be replaced
    c = 0           # index of said queue
    cluster = [center]
    while c < len(Q):   #until end of queue
        node = Q[c]     #get queue element
        c += 1
        for i in list(graph.adj[node]):     # for loop over neighbours
            if i not in cluster and distance(center,i) < t: #add to cluster if distance is less then t
                cluster.append(i)
                Q.append(i)
    return cluster



#create entire clustering from graph
def stoC(graph, t):
    G2 = G.copy()   # get copy of Graph, dont want to modify original
    clustering = [] # list of clusters
    while len(G2.nodes) > 0:    # we remove nodes that we put in clusters, keep going until all is in a cluster
        node = pickNode(G2)
        cluster = stoQuery(G2, t, node)
        G2.remove_nodes_from(cluster)   # remove nodes
        clustering.append(cluster)      # add cluster to list
    return clustering

File: clusters.ts
import { MailGraph } from "./mailGraph"

class clusters { //class to hold clusters

    stoQuery(graph: MailGraph, nodes: number[], t: number, center: number) { //creates single cluster from center
        let Q: number[] = [center]                          // queue of nodes to pass through
        let cluster: number[] = [center]                    // cluster to return
        while (Q.length > 0){
            let node: number = Q.shift() as number                           // pop queue
            for (let employee of graph.neighbours(node)){                    // add neighbours that are not yet in and within dist
                if (!cluster.includes(employee) && !Q.includes(employee) && nodes.includes(employee) && graph.distance(center, employee) > t) {
                    cluster.push(employee)
                    Q.push(employee)
                }
            }
        }
        return cluster
    }

    pickNode(nodes: number[]){ //get random element from the nodes
        return nodes[Math.floor(Math.random() * nodes.length)];
    }

    stoC(graph: MailGraph, t: number){
        let nodes: number[] = graph.copynodes()
        let clustering: number[][] = []
        while (nodes.length > 0) {                                 // as long as there are unclustered
            let node: number = this.pickNode(nodes)
            let cluster: number[] = this.stoQuery(graph, nodes, t, node)  //get cluster
            nodes = nodes.filter( x => !cluster.includes(x))        // remove from nodes
            clustering.push(cluster)
        }
        return clustering
    }



    clusterLinks(graph: MailGraph, clustering: number[][]){
        let links: number[][] = []
        for (let i = 0; i < clustering.length; i++){
            let link: number[] = Array(i).fill(-1)
            for (let j = i+1; j < clustering.length; j++){
                link.push(graph.clusterDist(clustering[i], clustering[j]))
            }
            links.push(link)
        }
        return links
    }
}

export default clusters